### PR TITLE
Handle /events in gateway mode

### PR DESCRIPTION
**Pull request by samrabelachew**
Rather than re-implement most of the existing Post workflow logic like I intended, to reduce code duplication, I extracted out an interface called `CommandExecutor` (admittedly could be better named) that the `VCSEventsController` uses. This lets users change the behavior behind running auto/comment commands. 

PR also creates a Gateway implementation of this interface. Requests that don't need processing are returned immediately and requests that do are sent through SNS to the worker.

### CODE REVIEW

File: server/controllers/events/gateway_command_executor.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func (g *GatewayCommandExecutor) ExecuteAutoplanCommand(request *http.Request, eventType models.PullRequestEventType, baseRepo models.Repo, headRepo models.Repo, pull models.PullRequest, user models.User, _ time.Time, _ logging.SimpleLogging) HttpResponse {
+	var err error
+	var respBody string
+	switch eventType {
+	case models.OpenedPullEvent, models.UpdatedPullEvent:
+		// If the pull request was opened or updated, we perform a pseudo-autoplan to determine if tf changes exist.
+		// If they exist, then we will forward request to the worker.
+		if hasTerraformChanges := g.CommandRunner.RunPseudoAutoplanCommand(baseRepo, headRepo, pull, user); hasTerraformChanges {
```
Comment (msarvar) [inline server/controllers/events/gateway_command_executor.go:46]: I don't think this worth running here. We should just put requests into the worker queue and move on. CommandRunner should figure out the rest.

File: server/controllers/events/gateway_command_executor.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			if err = g.SendToWorker(request); err == nil {
```
Comment (msarvar) [inline server/controllers/events/gateway_command_executor.go:47]: I think it reads better and follows golang convention to check for existence of the error. And deal with it right away

File: server/controllers/events/gateway_command_executor.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+type GatewayCommandExecutor struct {
+	SNSWriter     sns.Writer
+	CommandRunner events.CommandRunner
+	Scope         tally.Scope
+	Logger        logging.SimpleLogging
+}
+
+func (g *GatewayCommandExecutor) ExecuteCommentCommand(request *http.Request, _ models.Repo, _ *models.Repo, _ *models.PullRequest, _ models.User, _ int, _ *events.CommentCommand, _ time.Time) HttpResponse {
+	err := g.SendToWorker(request)
+	if err != nil {
```
Comment (Aayyush) [inline server/controllers/events/gateway_command_executor.go:24]: nit:
```suggestion
	if err := g.SendToWorker(request); err != nil {
```

File: server/controllers/events/gateway_command_executor.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	return HttpResponse{
+		body: respBody,
+	}
+}
+
+func (g *GatewayCommandExecutor) SendToWorker(r *http.Request) error {
+	buffer := bytes.NewBuffer([]byte{})
+	if err := r.Write(buffer); err != nil {
+		g.Scope.SubScope("send").Counter("failure").Inc(1)
+		return errors.Wrap(err, "Marshalling gateway request to buffer")
```
Comment (Aayyush) [inline server/controllers/events/gateway_command_executor.go:77]: nit: we use lowercase letters when wrapping errors. ([Style Guide](https://github.com/lyft/atlantis-private/blob/master/docs/internal/team_practices/style_guide.rst))
```suggestion
		return errors.Wrap(err, "marshalling gateway request to buffer")
```

File: server/controllers/events/gateway_command_executor.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	if err := g.SNSWriter.Write(buffer.Bytes()); err != nil {
+		g.Scope.SubScope("send").Counter("failure").Inc(1)
+		return errors.Wrap(err, "Writing gateway message to sns topic")
```
Comment (Aayyush) [inline server/controllers/events/gateway_command_executor.go:81]:
```suggestion
		return errors.Wrap(err, "writing gateway message to sns topic")
```

**Review by Aayyush — COMMENTED** [review]
Great job 🎉! Left a couple of nits

File: server/controllers/events/gateway_command_executor.go (hunk 1 above)
Comment (msarvar) [reply server/controllers/events/gateway_command_executor.go:46]: Ignore the comment, this was part of the requirements in the tech spec.

**Pull request closed without merging**